ISPN-12224 Cluster in a confusing state after restarted from graceful…

jabolina opened the pull request:
… shutdown - no hint for waiting on complete restarted

https://issues.redhat.com/browse/ISPN-12224

* Follow the partition policy when missing members after graceful shutdown;
* Only clear persistent state after successfully recovering after a graceful shutdown;

wburns reviewed (commented):
Just some test cleanup stuff.

wfink commented:
The fix prevent from loose the state if the node is started twice (i.e. after fail).
It will keep the cluster in DEGRADE state, but with the default settings, no partition-handling DENY_READ-WRITE enabled the cluster nodes are available for clients and the state is inconsistent if caches are modified.
So there is still no complete solution

jabolina commented:
I'll apply the suggestions, but I'll put this PR on hold until we discuss it further.

jabolina commented:
Rebased. Since most of the issues were fixed on #10479, I am leaving only the change I had introduced for the partition handling. Instead of always starting with `AVAILABLE`, it might start with `DEGRADED` and then move to `AVAILABLE` depending on the configuration.

The fixes for waiting for the cluster to restore and log about it are included on #10479. This one might be worth integrating, but it doesn't affect the behavior introduced by #10379 after the shutdown. We can see it as an additional safeguard.

wburns commented:
@jabolina Are you sure none of the test failures are releated? There are a few that seem to be related to nodes leaving the cluster with partition handling enabled.

jabolina commented:
@wburns I don't think so, or at least they don't show easily. I ran the core suite multiple times, and they didn't happen again.

wburns commented:
Looking closer the others do seem unrelated
